### PR TITLE
gcoap/fileserver: add file and directory creation and deletion

### PR DESCRIPTION
**Pull request by fabian18**
### Contribution description

With this PR I added write support for the `gcoap` fileserver, using `POST`, `PUT` and `DELETE` methods.

~`POST`: used to create files and directories~
`PUT` : creates files or directories or updates files in a random access fashion
`DELETE`: delete files and directories

CoAP options `If-Match` and `If-None-Match` are added and handled in `PUT` requests.

Somehow directory deletion does not work for me. I end up in `littlefs2/lfs.c:1457`.
```C
// check if we fit
lfs_size_t dsize = lfs_tag_dsize(tag);
if (commit->off + dsize > commit->end) {
    return LFS_ERR_NOSPC;
}
```
~I don´t know why. Maybe it works for you.~
Works with #18141

### Testing procedure

Flash `examples/gcoap_fileserver` to a board with Ethernet support, connect with your PC and use a CoAP client to upload some files to the board´s VFS.

<details>
<summary>RIOT shell</summary>

```
> vfs ls /nvm0
2022-05-24 11:53:27,139 # vfs ls /nvm0
2022-05-24 11:53:27,144 # ./
2022-05-24 11:53:27,144 # ../
2022-05-24 11:53:27,163 # anotherdir/
2022-05-24 11:53:27,189 # cacert.der    425 B
2022-05-24 11:53:27,221 # cakey.der     121 B
2022-05-24 11:53:27,260 # capubkey.der  91 B
2022-05-24 11:53:27,263 # total 3 files
vfs ls /nvm0
2022-05-24 11:53:47,834 # vfs ls /nvm0
2022-05-24 11:53:47,839 # ./
2022-05-24 11:53:47,840 # ../
2022-05-24 11:53:47,854 # RIOT.txt      1402 B
2022-05-24 11:53:47,877 # anotherdir/
2022-05-24 11:53:47,908 # cacert.der    425 B
2022-05-24 11:53:47,944 # cakey.der     121 B
2022-05-24 11:53:47,986 # capubkey.der  91 B
2022-05-24 11:53:47,987 # total 4 files
> vfs ls /nvm0
2022-05-24 11:55:30,836 # vfs ls /nvm0
2022-05-24 11:55:30,843 # ./
2022-05-24 11:55:30,843 # ../
2022-05-24 11:55:30,864 # anotherdir/
2022-05-24 11:55:30,892 # cacert.der    425 B
2022-05-24 11:55:30,926 # cakey.der     121 B
2022-05-24 11:55:30,962 # capubkey.der  91 B
2022-05-24 11:55:31,007 # tmp/
2022-05-24 11:55:31,008 # total 3 files
```
</details>

I test with `libcoap` example client. and `CFLAGS += -DCONFIG_GNRC_IPV6_NIB_ARSM=1`

<details>
<summary>Host shell</summary>

```
$ coap-client -m put -O 27,0x00 -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/RIOT.txt -f ~/RIOT.txt 
$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/RIOT.txt
                          ZZZZZZ
                        ZZZZZZZZZZZZ
                      ZZZZZZZZZZZZZZZZ
                     ZZZZZZZ     ZZZZZZ
                    ZZZZZZ        ZZZZZ
                    ZZZZZ          ZZZZ
                    ZZZZ           ZZZZZ
                    ZZZZ           ZZZZ
                    ZZZZ          ZZZZZ
                    ZZZZ        ZZZZZZ
                    ZZZZ     ZZZZZZZZ       777        7777       7777777777
              ZZ    ZZZZ   ZZZZZZZZ         777      77777777    77777777777
          ZZZZZZZ   ZZZZ  ZZZZZZZ           777     7777  7777       777
        ZZZZZZZZZ   ZZZZ    Z               777     777    777       777
       ZZZZZZ       ZZZZ                    777     777    777       777
      ZZZZZ         ZZZZ                    777     777    777       777
     ZZZZZ          ZZZZZ    ZZZZ           777     777    777       777
     ZZZZ           ZZZZZ    ZZZZZ          777     777    777       777
     ZZZZ           ZZZZZ     ZZZZZ         777     777    777       777
     ZZZZ           ZZZZ       ZZZZZ        777     777    777       777
     ZZZZZ         ZZZZZ        ZZZZZ       777     777    777       777
      ZZZZZZ     ZZZZZZ          ZZZZZ      777     7777777777       777
       ZZZZZZZZZZZZZZZ            ZZZZ      777      77777777        777
         ZZZZZZZZZZZ               Z
            ZZZZZ 

$ coap-client -m delete -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/RIOT.txt
$ coap-client -m put -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/tmp/
$ coap-client -m delete -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/tmp/
```

</details>

### CODE REVIEW

File: sys/net/application_layer/gcoap/fileserver.c
Comment (miri64) [inline sys/net/application_layer/gcoap/fileserver.c:40]: Were did this disappear to.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (miri64) [inline sys/net/application_layer/gcoap/fileserver.c:42]: IIRC there was a problem with flag fields (might be C++ or XCode related, but not sure), could be that I am cargo culting here. However, given that this is at max 8-bit long for now, this should go to the end of the sturct in any case to avoid padding.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:40]: Block2 is now only handled in the GET method handlers. Now that other methods are supported, I did not see a need to reserve memory for that in `struct requestdata`.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:45]: If there is a problem with bit fields, the CI will show this, right? Then I can of course turn this into flags.
I only know that you should not rely on the order of bits in bit fields.

Comment (miri64) [reply sys/net/application_layer/gcoap/fileserver.c:45]: > If there is a problem with bit fields, the CI will show this, right?

Not everything is tested, especially not XCode, as MacOSX support is currently broken for native (and we do not have a MacOSX worker). Not sure if there is a C++ application that uses this module.

**Comment by chrysn** [comment]
From a CoAP point of view, I think that PUT should be used for file creation -- this has not only better properties (w/rt idempotency) but is also aligned with what WebDAV does. (POST would have a place if you want to create a file whose name you don't care about, but I don't think that that is a common use case).

File: sys/net/application_layer/gcoap/fileserver.c
Comment (chrysn) [reply sys/net/application_layer/gcoap/fileserver.c:45]: The only problem I know of with bit fields is that they don't work like one might expect *when using the struct in type punning with a buffer* -- ie. you can't do `union { struct { bool ver: 2; bool mtype: 2; ...} parsed; uint8_t[4] for_network };` but for regular use I think they're fine.

Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:45]: `USEMODULE+="gcoap_fileserver gnrc_sock_udp gnrc_ipv6" make BOARD=nucleo-f767zi` compiles

Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:45]: for `examples/riot_and_cpp`

Comment (miri64) [reply sys/net/application_layer/gcoap/fileserver.c:45]: > The only problem I know of with bit fields is that they don't work like one might expect _when using the struct in type punning with a buffer_ -- ie. you can't do `union { struct { bool ver: 2; bool mtype: 2; ...} parsed; uint8_t[4] for_network };` but for regular use I think they're fine.

Then my cargo cult might stem from that, because I think I got hurt way back when I tried to send them over the wire :D.

Comment (miri64) [reply sys/net/application_layer/gcoap/fileserver.c:45]: I consider this discussion resolved then :-).

**Comment by fabian18** [comment]
> PUT should be used for file creation
Indeed PUT can also be used for resource creation according to RFC7252. The handlers for POST and PUT anyways show some code duplication. I should be able to reduce it to just PUT.

**Comment by chrysn** [comment]
>  should be able to reduce it to just PUT.

I think that makes things simper.

It also aligns it with [the CoAP file service draft](https://www.ietf.org/archive/id/draft-amsuess-core-coap-kitchensink-00.html#name-coap). I feel obliged to point out that that is *very* drafty and has not even been properly discussed. If you want to consider that text for here, I'd rather hope for any disagreements between the text and the implementation would lead to a discussion on what good behavior is. (For example, this changes files non-atomically; maybe the text should rather not make promises unless one or the other behavior is advertised).

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:127]:
```suggestion
              code >> 5, code & 0x1f);
```

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:166]: This could need a comment explaining what the if_match does.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:386]:
```suggestion
    switch (coap_get_code(pdu)) {
    case COAP_METHOD_GET:
        return _get_file(pdu, buf, len, request, resource);
    case COAP_METHOD_PUT:
        return _put_file(pdu, buf, len, request, resource);
    case COAP_METHOD_DELETE:
        return _delete_file(pdu, buf, len, request, resource);
    default:
        return gcoap_fileserver_error_handler(pdu, buf, len, COAP_CODE_METHOD_NOT_ALLOWED);
    }
```

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:254]: This needs some explaining.
Why not use `strrchr(request->namebuf, '/')`?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:342]: label names are typically lower case in RIOT.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:275]: Why the new scope?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:342]: A second label saves you from writing `vfs_close(fd);` on each error condition with the file open.

```suggestion
close_on_error:
    vfs_close(fd);
CLEAN_ON_ERROR:
```

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:360]: So this is to only delete the file if the etag matches, right?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:456]: Should creating the directory be a separate request or should it happen automatically when a file is PUT into a directory that does not exist yet?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:479]: Is there the possibility to signal recursive deletion?
Or is the CoAP fileserver API fundamentally low-level?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:528]: Why the new scope?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:166]: I am going to add an explanation what `If-Match` and `If-None-Match` are supposed to be used for in this example, in `fileserver.h`.

The content of an `If-Match` option, sent by a client, is a previously received Etag. What it expresses is the following: The client wants to make sure that the resource it wants to update has not been updated by someone else, since the client had last received that Etag. The `If-Match` option can have 0 length which basically means: If the resource exists, I want my request to be processed. If the `If-Match` option does not match, the request failed to meet its precondition.

The `If-None-Match` option has length 0. It expresses that a client wants to create a resource only if it has not been created already.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:254]: I was not aware of `strrchr()`. But it does what you think it does: move the file component one char to the right and put a '.' in front of it.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:342]: nice

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:360]: yes

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:456]: According to the comments in the doc, the directory should be created implicitly. But I think it is simpler if we assume that the directory must be created in advance because if something goes wrong during the file creation, it would be more effort for us to revert the created directories.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:479]: Personally I am not interested in recursive deletion. Only empty directories can be deleted with `vfs_rmdir`,
If we want to implement recursive deletion later, we could define that a client must append a URI query `DELETE: coap://server:port/path/to/resource?recursive=1`. But this is just a first idea.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:528]: to limit the scope of inside defined variables. Do you not think that this is worth the scope overhead?

File: sys/net/application_layer/gcoap/fileserver.c
Comment (chrysn) [reply sys/net/application_layer/gcoap/fileserver.c:435]: I wrote that under the influence of git (helluva drug ;-) ) -- but I've come to think that explicit creation is cleaner (also because it's symmetric with that we don't delete directories when deleting the last files in them); that it's easier to to implement seems to confirm that that's the better direction.

coap-kitchensink-01 will recommend that an empty PUT (with no content-format) be used to create a directory. (It'd be an error to have a payload in that.)

File: sys/net/application_layer/gcoap/fileserver.c
Comment (chrysn) [reply sys/net/application_layer/gcoap/fileserver.c:458]: I suggest that recursive is the default. The UNIX behavior can be easily emulated with ETags (just pick a special ETag for empty directories on the server side, and insist that the client send an `If-Match` if a directory would be removed recursively), whereas the other direction is cumbersome to emulate by the client.

Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:458]: Do I understand this right?
If the client sends a `DELETE` on a directory, without an `If-Match`, the server should recursively delete the directory and all of its sub content.
If the client sends a `DELETE` on a directory, with an `If-Match` that contains the special Etag for empty directories, the server only deletes the directory if it is empty.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:45]: Some comments would be good here, it's not obvious when those flags would be set or what `if_match` is

**Review by benpicco — COMMENTED** [review]
If @chrysn has no objections to the server logic, this should be fine.

File: sys/net/application_layer/gcoap/fileserver.c
Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:256]: Should be safe because `vfs_unlink()` at the bottom is guarded with `if (create) { }` so only the temporary file would be deleted.

**Comment by fabian18** [comment]
> If @chrysn has no objections to the server logic/API, this should be fine.

I still need to implement the recursive DELETE. But I was not sure if I understood it right.
The coap fileserver draft could benefit from an example.

**Comment by benpicco** [comment]
Recursive delete might be generally useful, so best implement it in `vfs_util`

File: sys/vfs_util/vfs_util.c
Comment (benpicco) [inline sys/vfs_util/vfs_util.c:247]: No need to be so dense here - and since ".." is const anyway, no need for the `strncmp`

```suggestion
            seg_len = strlen(entry.d_name);
            if (!strcmp(entry.d_name, "..")) {
```

File: sys/vfs_util/vfs_util.c
Comment (benpicco) [inline sys/vfs_util/vfs_util.c:250]: That's just 
```suggestion
            root_len = strlen(path_buf);
            if (root_len + seg_len >= max_size) {
```

**Review by benpicco — COMMENTED** [review]
Looks pretty good!
No need to wait for an approval to rebase btw to fix the merge conflicts :wink: 

Just one more thing: Can you make the write support a compile-time conditional (`gcoap_fileserver_read_only`? `gcoap_fileserver_write`?) so read-only applications can save a few bytes and don't have to worry about possible security implications.

File: sys/vfs_util/vfs_util.c
Comment (fabian18) [reply sys/vfs_util/vfs_util.c]: I had `root` in mind when I chose `r`. It always points to the end, where the segment is appended.
I changed the names `r` --> `p_write` for write pointer and `seg` --> `p_read` for read pointer.

**Comment by fabian18** [comment]
> Can you make the write support a compile-time conditional

I am going to do that. I would add pseudomodules `gcoap_fileserver_put` that enables code in the `PUT` handler and `gcoap_fileserver_delete` that enables code in the `DELETE` handler.

**Comment by fabian18** [comment]
Let me know if I can squash

**Comment by benpicco** [comment]
Please squash!

File: sys/net/application_layer/gcoap/fileserver.c
Comment (benpicco) [inline sys/net/application_layer/gcoap/fileserver.c:39]: If a client is supposed to use this, shouldn't this be made available in a header?

Comment (miri64) [reply sys/net/application_layer/gcoap/fileserver.c:39]: Also (sorry for jumping in so late, if this was discussed before), why is an ETag the signifier for this? Usually, with a REST API, this is done with some specific parameters (either as a query parameter or e.g. as a CBOR/JSON object).

Comment (miri64) [reply sys/net/application_layer/gcoap/fileserver.c:39]: Ah, I saw that this was proposed by @chrysn in https://github.com/RIOT-OS/RIOT/pull/18133#discussion_r884165690. But still, I am a bit confused about this

Comment (fabian18) [reply sys/net/application_layer/gcoap/fileserver.c:39]: > But still, I am a bit confused about this

The content of an `If-Match` option is an Etag.
Simply formulated: If the client sends that special Etag, it expects the directory to be empty. And if so, the request is processed, which means it is `DELETE`ed. My first idea was indeed a query parameter, but I think the `If-Match` option works as well. Does this help or what exactly is unclear?

**Review by benpicco — APPROVED** [review]
Works like a charm!
For testing with `coap-client` I had to ignore the URI_HOST option as it would send it and it's a critical option - but we aren't doing vhosting, so any hostname will do.

Since it's just a two line change I've taken the liberty to just push that to your branch.

**Comment by fabian18** [comment]
Awesome, Thank You!

**Pull request merged**
